### ISSUE TITLE
backup/restore as part of ci/cd process?

### ISSUE THREAD

**Issue by jyjohnson** [sethkor/connect-backup#26]
Hi - first, thanks for creating this super useful tool.

I'm curious if it can be extended to support a traditional deployment model. For example, I would like to create Contact Flow(s) in a 'dev' AWS account ('dev' Connect instance), test/qa, and if approved, deploy the update Contact Flow(s) to another AWS account ('prod' Connect instance). I see the json elements use the ARN which includes the AWS account # and the Connect instance ID. Does the tool change these when doing a restore to another account/instance? If not could that be added?

**Comment by sethkor** [comment]
Hey @jyjohnson  , apologies for the delayed response, I've been taking a bit of a holiday.  I'll take a look at your request but it sounds reasonable.  Give me a couple of days to look into it.

**Comment by sethkor** [comment]
Ah so a little more work @jyjohnson as ARNS are littered everywhere, and theres no consistent pattern.  Also I should probably consider region as something to be accommodated by this use case.  Flows are nested too with unique id's (whispers etc).  Still I think it's a worthwhile feature request but it's going to take a little longer to do.

**Comment by sethkor** [comment]
Right-o.  So the big challenge here is objects (flows, queues, prompts etc) all have unique id.  They look something like `b20f58bb-e5cc-4618-9e7b-3ee3d446539e`
These id are uniquely generated when the object is created.  An ARN in connect has the unique id as the suffix of the ARN.
These id are then referenced in contact flows and flows can be nested too.  Also there is no API call right now to create prompts or queues (and perhaps some more things I need to dig into).
So in order to restore a flow to a new connect instance (essentially the prefix of the ARN) you have to ensure all dependencies are created first, then update references to them in the contact flow.  The referencing looks like it can be done in code via a name but you will need to manually create some of the objects (queues, prompts) first and ensure the exact same name is used consistently between your connect instances.  For example if you have an announcement named "Say Hello World" you will need to ensure all of your instances have this announcement manually created with this exact name in order for a restoration of a contact flow to work properly. Same goes for queues.

Given the above AWS Connect limitations, is this feature still useful to you @jyjohnson ?

**Comment by jyjohnson** [comment]
Howdy. I think assuming all dependencies are in place is "ok". I haven't looked at the API methods but could it check that all named resources exist in the target account/instance? Overall I think this would be useful. Thanks.

**Comment by jasondearingcnx** [comment]
Hello.  Thanks for making such a useful tool.  Does connect-backup currently support SSO for aws profiles?  I'm getting an error, but only when I try to use a profile I've signed in as using "aws configure sso":
2021/05/25 14:52:22 NoCredentialProviders: no valid providers in chain
caused by: EnvAccessKeyNotFound: failed to find credentials in the environment.
SharedCredsLoad: failed to load profile, <mySSOProfileName>.
EC2RoleRequestError: no EC2 instance role found
caused by: RequestError: send request failed

**Comment by sethkor** [comment]
Started working on this given multiple requests.  It's a lot of work and will involve a few smaller features along the way.  Stay tuned.
